### PR TITLE
VXFM-2097 Be able to input the root creds for FlexOS gateway during discovery

### PR DESCRIPTION
**Pull request by aimers1975**
Update inventory script to use gateway OS password and fail discovery
if we can't ssh using those credentials.

Should be merged with:  
https://github.com/dell-asm/asm-core/pull/4032
https://github.com/dell-asm/asm-deployer/pull/2539

### CODE REVIEW

File: bin/discovery.rb [hunk 1]
```diff
@@ -13,6 +14,9 @@
   opt :port, "ScaleIO gateway port", :default => 443
   opt :username, "ScaleIO gateway username", :type => :string, :required => true
   opt :password, "ScaleIO gateway password", :type => :string, :default => ENV["PASSWORD"]
+  opt :gatewayosusername, "", :type => :string, :required => true
+  opt :gatewayospasswordid, "Not used"
+  opt :gatewayospassword, "", :type => :string, :default => ENV["GATEWAYOSPASSWORD"]
```
Comment (gavin-scott) [inline bin/discovery.rb]: Can we call these just e.g. `os_username`, `os_password_id`, `os_password`?

File: bin/discovery.rb [hunk 2]
```diff
@@ -83,6 +88,17 @@ def collect_scaleio_facts
   facts
 end
 
+def check_ssh_connection
+  puts "Attempting to ssh to verify root credentials."
+  begin
+    result = ASM::Util.execute_script_via_ssh(@opts[:server], @opts[:gatewayosusername], @opts[:gatewayospassword], "ls", "")
```
Comment (gavin-scott) [inline bin/discovery.rb]: Why pass `""` as an argument to `ls`? To make sure that it doesn't list any actually directory?

Also, this method seems like it always returns a truthy value so saving and checking `result` doesn't make sense. I'm assuming you are relying on an exception being thrown as the failure case?

Comment (aimers1975) [reply bin/discovery.rb]: The method call seemed to fail when I didn't pass that param...I believe...I can got back and retry again, but I think originally I didn't provide any arguments and something complained...I can sped some time going back and looking at it...

File: bin/discovery.rb (hunk 1 above)
Comment (aimers1975) [reply bin/discovery.rb]: I was keeping them consistent with what they were being called on java side...I can see about changing it...if you don't want it that way.

File: bin/discovery.rb [hunk 3]
```diff
@@ -13,6 +14,9 @@
   opt :port, "ScaleIO gateway port", :default => 443
   opt :username, "ScaleIO gateway username", :type => :string, :required => true
   opt :password, "ScaleIO gateway password", :type => :string, :default => ENV["PASSWORD"]
+  opt :osusername, "", :type => :string, :required => true
+  opt :ospasswordid, "Not used"
+  opt :ospassword, "", :type => :string, :default => ENV["OSPASSWORD"]
```
Comment (gavin-scott) [inline bin/discovery.rb]: Can we do `os_username` and `os_password`?

File: bin/discovery.rb [hunk 4]
```diff
@@ -179,7 +195,7 @@ def scaleio_cookie
       results ||= {}
       scaleio_cache = "/opt/Dell/ASM/cache"
       Dir.mkdir(scaleio_cache) unless Dir.exists? scaleio_cache
-      file_path = File.join(scaleio_cache, "#{opts[:server]}.json")
+      file_path = File.join(scaleio_cache, "#{@opts[:server]}.json")
```
Comment (gavin-scott) [inline bin/discovery.rb]: This is interesting. Did the old code work also or was it writing an invalid filename? Your change makes it more consistent, just curious.

Comment (aimers1975) [reply bin/discovery.rb]: when I found this, I got an exception from it...so I am baffled a little...

**Comment by aimers1975** [comment]
@gavin-scott Did you have any other feedback on this PR?

**Pull request merged**